RNA-seq/06-openpbta_heatmap.Rmd

```diff
@@ -0,0 +1,385 @@
+
+### Heatmap itself!
+
+Okay, now we're ready to make a heatmap complete with annotation bars.
+
+```{r make_heatmap}
+Heatmap(zscores_mat,
+        # The distance metric used for clustering the rows
+        # This is different from the default (Euclidean)
+        clustering_distance_rows = "pearson",
+        # Linkage method for row clustering
+        # This is different from the default (complete)
+        clustering_method_rows = "average",
+        # Distance metric for columns
+        clustering_distance_columns = "pearson",
+        # Linkage for columns
+        clustering_method_columns = "average",
+        show_row_names = FALSE,
+        show_column_names = FALSE,
+        # Add annotation bars to the top of the 
+        top_annotation = column_annotation,
+        # This will be used as the label for the color bar
+        # of the cells of the heatmap itself
+        name = "z-score")
+```
+
+We have to create the heatmap _twice_ -- once for display in the notebook and once to save to a PNG file.
+
+```{r save_heatmap_png, live = TRUE}
+# Open PNG plot device
+png(filename = heatmap_file, 
+    width = 11, 
+    height = 7, 
+    units = "in", 
+    res = 300)
+# Heatmap, again!
+Heatmap(zscores_mat,
+        clustering_distance_rows = "pearson",
+        clustering_method_rows = "average",
+        clustering_distance_columns = "pearson",
+        clustering_method_columns = "average",
+        show_row_names = FALSE,
+        show_column_names = FALSE,
+        top_annotation = column_annotation,
+        name = "z-score",
+        use_raster = FALSE) # higher resolution for output (be careful with PDF output!)
+# Shut down current graphics device
+dev.off()
+```
+
+### Notes on clustering
+
+You probably noticed that there were some arguments that we defined that were related to the clustering of samples and genes when we build the heatmaps.
+The kind of clustering we performed was [hierarchical clustering](https://en.wikipedia.org/wiki/Hierarchical_clustering).
+This is an _agglomerative_ method of clustering - each sample starts in its own cluster and then at each step of the algorithm the two most similar clusters are joined until there's only a single cluster left.
+
+We defined the distance between samples as 1 minus the correlation between them, and then averaged samples when they were merged into a group.
+The method we're using is also known as [UPGMA (unweighted pair group method with arithmetic mean)](https://en.wikipedia.org/wiki/UPGMA).
+
+It is important to note that the linear arrangement of samples in the the tree is somewhat arbitrary, so we have to be careful not to overintepret it.
+We can rotate around any branch of the tree in our visualization with no change to the tree topology itself.
+In fact, the samples at opposite ends of the heatmap could be right next to one another with the right set of rotations!
```

Comment:
Do you want to talk about branch lengths here? Caveat: I don't know much about dendrogram defaults in `ComplexHeatmap`.

Reply:
No, I do not want to talk about branch lengths.